executor/infoschema_reader.go

```diff
@@ -1760,3 +1767,193 @@ func (e *memtableRetriever) setDataForStatementsSummary(ctx sessionctx.Context,
 	}
 	return nil
 }
+
+// TiFlashSystemTableRetriever is used to read system table from tiflash.
+type TiFlashSystemTableRetriever struct {
+	dummyCloser
+	table         *model.TableInfo
+	outputCols    []*model.ColumnInfo
+	instanceCount int
+	instanceIdx   int
+	instanceInfos []tiflashInstanceInfo
+	rowIdx        int
+	retrieved     bool
+	initialized   bool
+	extractor     *plannercore.TiFlashSystemTableExtractor
+}
+
+func (e *TiFlashSystemTableRetriever) retrieve(ctx context.Context, sctx sessionctx.Context) ([][]types.Datum, error) {
+	if e.extractor.SkipRequest || e.retrieved {
+		return nil, nil
+	}
+	if !e.initialized {
+		err := e.initialize(sctx, e.extractor.TiFlashInstances)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if e.instanceCount == 0 || e.instanceIdx >= e.instanceCount {
+		e.retrieved = true
+		return nil, nil
+	}
+
+	for {
+		rows, err := e.dataForTiFlashSystemTables(sctx, e.extractor.TiDBDatabases, e.extractor.TiDBTables)
+		if err != nil {
+			return nil, err
+		}
+		if len(rows) > 0 || e.instanceIdx >= e.instanceCount {
+			return rows, nil
+		}
+	}
+}
+
+type tiflashInstanceInfo struct {
+	id  string
+	url string
+}
+
+func (e *TiFlashSystemTableRetriever) initialize(sctx sessionctx.Context, tiflashInstances set.StringSet) error {
+	store := sctx.GetStore()
+	if etcd, ok := store.(tikv.EtcdBackend); ok {
+		if addrs := etcd.EtcdAddrs(); addrs != nil {
```

Comment:
How about use `domain.GetDomain(sctx).GetEtcdClient()` to get the etcd client?

Reply:
Fixed.